Content.Shared/Store/ListingPrototype.cs

```diff
@@ -77,6 +78,19 @@ public partial class ListingData : IEquatable<ListingData>, ICloneable
     [DataField("productAction", customTypeSerializer: typeof(PrototypeIdSerializer<EntityPrototype>))]
     public string? ProductAction;
 
+    /// <summary>
+    ///     The listing ID of the related upgrade listing. Can be used to link a <see cref="ProductAction"/> to an
+    ///         upgrade or to use standalone as an upgrade
+    /// </summary>
+    [DataField]
+    public ProtoId<ListingPrototype>? ProductUpgradeID;
+
+    /// <summary>
+    ///     Keeps track of the current action entity this is tied to, for action upgrades
+    /// </summary>
+    [NonSerialized]
+    public EntityUid? ProductActionEntity;
+
```

Comment:
This is particularly annoying because it needs to be a datafield and a NetEntity as this is just going to break things for persistence so it really needs fixing.

Also you missed updating the Equals method.